### PR TITLE
Migrate PenaltyKickTactic to Use PenaltyKickFSM

### PR DESCRIPTION
**Pull request by itsarune**
### Description

- Migrated PenaltyKickTactic to use the FSM framework
- evaluatePenaltyShot was rewritten slightly to make approach to keeper less finicky
- made it compliant with SSL rules (coroutine version didn't obey excessive dribbling rules)

output

### Testing Done

- existing tests were moved and also now pass

### Length Justification

- a lot of this code are tests that already existed but were moved to a new location

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by akhilveeraghanta — COMMENTED** [review]
Awesome work :fire:! Do you plan to add simulated tactic tests for the penalty kick tactic as part of this PR?

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+#include "software/ai/hl/stp/tactic/tactic.h"
+#include "software/geom/algorithms/closest_point.h"
+#include "software/geom/algorithms/intersection.h"
+
+struct PenaltyKickTacticFSM
+{
+    class ApproachBallState;
+    class ApproachKeeperState;
+    class InitialState;
+    class ShootState;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: seems like only InitialState is being used?

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h [hunk 2]
```diff
@@ -0,0 +1,299 @@
+#pragma once
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: can we move the penalty kick files to a `tactic/penalty_kick` folder?

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+         * @param event  PenaltyKickTacticFSM::Update
+         */
+        const auto ball_kicked = [](auto event) {
+            return event.common.world.ball().hasBallBeenKicked(shot_angle,
+                                                               PENALTY_KICK_SHOT_SPEED);
+        };
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest state
+            *initial_s + update_e / initialize_start_keeper = approach_keeper_s,
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: I think you can start the FSM in the approach_keeper_s state

**Review by jonathanlew — COMMENTED** [review]
good job, it'd be good to see some simulated tactic tests

File: src/software/ai/hl/stp/tactic/dribble/dribble_fsm.h [hunk 4]
```diff
@@ -34,7 +34,7 @@ struct DribbleFSM
     static constexpr double DRIBBLE_KICK_SPEED = 0.05;
     // Maximum distance to continuously dribble the ball, slightly conservative to not
     // break the 1 meter rule
-    static constexpr double MAX_CONTINUOUS_DRIBBLING_DISTANCE = 0.9;
+    static constexpr double MAX_CONTINUOUS_DRIBBLING_DISTANCE = 0.87;
```
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/dribble/dribble_fsm.h]: Yes that's why I changed the constant

https://robocup-ssl.github.io/ssl-rules/sslrules.html#_penalty_kick

> The restrictions defined for scoring goals, including the ball height limit of 0.15 meters, do not apply here. Other rules like the excessive dribbling limitation for example do.

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                               penalty_start_approaching_goalie)
+                                  .toSeconds();
+                const Point next_robot_position =
+                    event.common.world.field().friendlyPenaltyMark() +
+                    Vector(approach_keeper_speed * time, 0);
+                current_robot_position = event.common.robot.position();
+                DribbleFSM::ControlParams control_params{
+                    .dribble_destination = std::optional<Point>(next_robot_position),
+                    .final_dribble_orientation = std::nullopt,
+                    .allow_excessive_dribbling = false,
```
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_penalty_kick

> The restrictions defined for scoring goals, including the ball height limit of 0.15 meters, do not apply here. Other rules like the excessive dribbling limitation for example do.

[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: oh nice catch

File: src/software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic_fsm.h [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        }
+    }
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        const auto approach_keeper_s = state<DribbleFSM>;
+        // const auto adjust_orientation_for_shot = state<DribbleFSM>;
+        // const auto initial_s         = state<InitialState>;
```
Comment (HaomiaoZ) [inline src/software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic_fsm.h]: Are those variables needed?

File: src/software/ai/hl/stp/tactic/penalty_kick/simulated_penalty_kick_tactic_test.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    protected :
+        void SetUp() override
+        {
+            setBallState(BallState(field().friendlyPenaltyMark(), Vector(0, 0)));
+            Point initial_position = field().friendlyPenaltyMark() + Vector(-0.1, 0);
+            addFriendlyRobots(TestUtil::createStationaryRobotStatesWithId({initial_position}));
+        }
+};
+
+// TEST_P(SimulatedPenaltyKickTacticTest, penalty_kick_test)
```
Comment (HaomiaoZ) [inline src/software/ai/hl/stp/tactic/penalty_kick/simulated_penalty_kick_tactic_test.cpp]: just wondering why `TEST_P` is commented out here.

**Review by HaomiaoZ — COMMENTED** [review]
I think generally everything looks good to me.  I just have some questions on the commented out code.

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+void PenaltyKickTactic::accept(TacticVisitor& visitor) const
+{
+    visitor.visit(*this);
+}
+
+Ball PenaltyKickTactic::getBall() const
+{
+    return this->ball;
+}
```
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.cpp]: removed

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
+/**
+ * Implementation of the PenaltyKickTactic
+ */
+#include "software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.h"
+
+PenaltyKickTactic::PenaltyKickTactic(const Ball& ball, bool loop_forever,
```
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.cpp]: I removed both, and made the constructor call to loop_forever to always be false

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.cpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+                                     const std::optional<Robot>& enemy_goalie)
+    : Tactic(loop_forever,
+             {RobotCapability::Move, RobotCapability::Dribble, RobotCapability::Kick}),
+      ball(ball),
+      enemy_goalie(enemy_goalie),
```
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic.cpp]: removed

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+                                                               PENALTY_KICK_SHOT_SPEED);
+        };
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest state
+            *initial_s + update_e / initialize_start_keeper = approach_keeper_s,
+            approach_keeper_s + update_e[!take_penalty_shot] / update_approach_keeper,
+            approach_keeper_s + update_e[take_penalty_shot] / adjust_orientation_for_shot,
+            approach_keeper_s = shoot_s, shoot_s + update_e[!ball_kicked] / shoot,
+            shoot_s + update_e[ball_kicked] = X);
```
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: updated

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h (hunk 1 above)
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: forgot to remove them from an earlier version

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h (hunk 2 above)
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: done

File: src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h (hunk 3 above)
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick_tactic/penalty_kick_tactic_fsm.h]: done, now some of the shared variables are initialized using a std::optional param

File: src/software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic_fsm.h (hunk 6 above)
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic_fsm.h]: no, this was code that I was playing around with while making PR changes, got rid of it now

File: src/software/ai/hl/stp/tactic/penalty_kick/simulated_penalty_kick_tactic_test.cpp (hunk 7 above)
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/penalty_kick/simulated_penalty_kick_tactic_test.cpp]: same thing as above, it's active now

File: src/software/gui/full_system/threaded_full_system_gui.cpp [hunk 12]
```diff
@@ -67,7 +67,7 @@ void ThreadedFullSystemGUI::createAndRunFullSystemGUI()
         world_draw_functions_buffer, ai_draw_functions_buffer, play_info_buffer,
         sensor_msg_buffer, view_area_buffer, worlds_received_per_second_buffer,
         primitives_sent_per_second_buffer, mutable_thunderbots_config);
-    full_system_gui->show();
+    full_system_gui->showFullScreen();
```
Comment (jonathanlew) [inline src/software/gui/full_system/threaded_full_system_gui.cpp]: this should be behind a command line arg or removed from this PR

File: src/software/ai/hl/stp/tactic/dribble/dribble_fsm.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/dribble/dribble_fsm.h]: it looks like tests pass with keeping 0.9, so might as well revert?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
just couple comments and it should be good to go from my end

File: src/software/gui/full_system/threaded_full_system_gui.cpp (hunk 12 above)
Comment (itsarune) [reply src/software/gui/full_system/threaded_full_system_gui.cpp]: removed

File: src/software/ai/hl/stp/tactic/dribble/dribble_fsm.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/dribble/dribble_fsm.h]: makes sense

**Review by akhilveeraghanta — APPROVED** [review]
well done! LGTM 💯

**Review by jonathanlew — APPROVED** [review]
Nice work!

**Pull request merged**